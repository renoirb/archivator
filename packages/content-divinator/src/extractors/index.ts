export * from './words'
export * from './summary'
