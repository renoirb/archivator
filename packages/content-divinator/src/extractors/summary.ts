import { SummaryRecordType, WordUsageMapType } from '../types'

/**
 * Summarizing content.
 *
 * For starters, what are the most used "keywords".
 * More things to come.
 *
 * ----
 *
 * See earlier implementation:
 *   v2.0.0, rework implementation:
 *     links:
 *       - name: sortedAndKeywords
 *         replacing: analyze
 *         replacedWith: summary
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L50-L62
 *
 * ----
 *
 * @package extractors
 *
 * @param wordsMap - Unique "word" where each value is its usage count
 * @param floor - In keywords grouping, what is the minimum number of times to qualify (default: 3)
 * @param max - In keywords grouping, how many items in the top keywords (default: 10)
 */
export const summary = (
  wordsMap: WordUsageMapType,
  floor: number = 3,
  max: number = 10,
): SummaryRecordType => {
  const zeroIndexSafeFloor = Number.isInteger(+floor) && +floor > 0 ? floor : 1
  const keywords: string[] = []
  let iter = 0
  // @ts-ignore
  for (const [word, count] of wordsMap.entries()) {
    if (iter < max && count > zeroIndexSafeFloor) {
      keywords.push(word)
    }
    iter++
  }

  const out: SummaryRecordType = { keywords }

  return out
}
