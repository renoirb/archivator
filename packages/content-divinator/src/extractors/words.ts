import {
  wordNormalizer,
  nonStopWordIsser,
  convertRecordHashMapToMap,
} from '../utils'
import type { WordUsageMapType } from '../types'

/**
 * Extract words and calculate usage frequency.
 *
 * @internal
 * @package extractors
 *
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L22-L46 replacing v1.0 implementation of extractWords}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L76-L94 replacing v1.0 implementation of analyze}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L63-L74 replacing v1.0 implementation of sort}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L18-L20 replacing v1.0 implementation of removePunctuation}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L13-L16 replacing v2.0 implementation of normalize as wordNormalizer}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L18-L48 replacing v2.0 implementation of extractWords as words}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L64-L77 replacing v2.0 implementation of sort}
 *
 * @param body - Text content, as a single string
 * @param stopWords - Words that should be ignored
 * @param locales - Locales tags to support for {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleLowerCase|String.prototype.toLocaleLowerCase} (default: `['en-ca']`)
 */
export const words = (
  body: string,
  stopWords: string[] = [],
  locales: string | string[] = ['en-CA'],
): WordUsageMapType => {
  const normalizeWord = wordNormalizer(locales)
  const isNonStopWord = nonStopWordIsser(stopWords)
  // We want to use Object.create(null) instead of Map
  // because we want to quickly increment values on an Hash-Map
  const words: Record<string, number> = Object.create(null)
  const foundOnce = new Set<string>()
  const text = body.split(' ')
  for (let i = 0; i < text.length; i++) {
    const word = normalizeWord(text[i])
    if (word && isNonStopWord(word)) {
      if (foundOnce.has(word)) {
        // This was the former way to check, before we had Map.
        // Leaving here for old memoriesâ€™ sake.
        if (Object.prototype.hasOwnProperty.call(words, word)) {
          words[word]++
        } else {
          words[word] = 2
        }
      } else {
        foundOnce.add(word)
      }
    }
  }

  const sorted = convertRecordHashMapToMap(words)

  return Object.seal(sorted)
}
