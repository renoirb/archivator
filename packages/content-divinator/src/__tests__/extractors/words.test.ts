/* eslint-env jest */

import { words, summary } from '../../extractors'

import { fixtures } from '..'

describe('extractors', () => {
  const stopWords = fixtures.loadJson<ReadonlyArray<string>>(
    'english-stop-words-alpha.json',
  )
  const textContent = fixtures.loadText('article-alpha.txt')

  test('Happy-Path', () => {
    const subject = words(textContent, [...stopWords])
    for (const w of stopWords) {
      expect(subject).not.toHaveProperty(w)
    }
    expect(subject).toMatchSnapshot()
  })

  test('When adding stopWords', () => {
    // We're removing 'to' as a stop word
    const textHashMap = words(textContent, [...stopWords, 'to'])
    // But 'to' should be present in the top 20 of the words used at least 10 times
    const summarized = summary(textHashMap, 10, 20)
    expect(summarized).toMatchSnapshot()
    expect(summarized.keywords).toMatchObject([
      // 'to',
      'api',
      'the',
      'data',
      'that',
      'in',
      'return',
    ])
  })
})
