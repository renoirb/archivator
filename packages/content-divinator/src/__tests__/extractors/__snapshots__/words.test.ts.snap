// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractors Happy-Path 1`] = `
Map {
  "to" => 58,
  "api" => 39,
  "the" => 37,
  "data" => 21,
  "that" => 17,
  "in" => 15,
  "return" => 13,
  "not" => 10,
  "grains" => 10,
  "you" => 8,
  "salt" => 8,
  "script" => 8,
  "def" => 8,
  "line" => 8,
  "ec2" => 7,
  "metadata" => 7,
  "if" => 7,
  "some" => 6,
  "cluster" => 6,
  "serr" => 6,
  "this" => 6,
  "openstack" => 6,
  "i" => 6,
  "for" => 6,
  "from" => 6,
  "get" => 5,
  "dreamcompute" => 5,
  "file" => 5,
  "path" => 5,
  "original" => 5,
  "read" => 5,
  "except" => 5,
  "all" => 4,
  "import" => 4,
  "information" => 4,
  "key" => 4,
  "hostname" => 4,
  "on" => 4,
  "what" => 4,
  "call_response_data" => 4,
  "else" => 4,
  "it" => 4,
  "aws" => 4,
  "name" => 3,
  "grainsget" => 3,
  "minion" => 3,
  "work" => 3,
  "will" => 3,
  "but" => 3,
  "available" => 3,
  "my" => 3,
  "s" => 3,
  "case" => 3,
  "saltnovalocal" => 3,
  "iad1" => 3,
  "instance" => 3,
  "make" => 3,
  "_call_awsurl" => 3,
  "keys_to_mute" => 3,
  "url" => 3,
  "dline" => 3,
  "publickeys" => 3,
  "was" => 3,
  "your" => 3,
  "grain" => 3,
  "python" => 3,
  "loginfocould" => 3,
  "latestmetadatapublicipv4info" => 2,
  "latestmetadatainstanceidinfo" => 2,
  "latestmetadatainstancetypeinfo" => 2,
  "latestmetadatalocalipv4info" => 2,
  "latestmetadataplacementinfo" => 2,
  "latestmetadataplacementavailabilityzoneinfo" => 2,
  "80" => 2,
  "latestmetadataramdiskidinfo" => 2,
  "latestmetadatareservationidinfo" => 2,
  "http" => 2,
  "nova" => 2,
  "within" => 2,
  "vda" => 2,
  "have" => 2,
  "more" => 2,
  "app1" => 2,
  "network_config" => 2,
  "folder" => 2,
  "_grains" => 2,
  "availability_zone" => 2,
  "logging" => 2,
  "log" => 2,
  "200" => 2,
  "httplib" => 2,
  "conn" => 2,
  "httplibhttpconnection169254169254" => 2,
  "timeout1" => 2,
  "connrequestget" => 2,
  "securitygroups" => 2,
  "service" => 2,
  "doesnt" => 2,
  "10101011" => 2,
  "203011311" => 2,
  "internal" => 2,
  "dreamcomputeopenstack" => 2,
  "following" => 2,
  "latestmetadata" => 2,
  "vm" => 2,
  "lightspeed" => 2,
  "placement" => 2,
  "each" => 2,
  "know" => 2,
  "sshrsa" => 2,
  "public" => 2,
  "function" => 2,
  "keys" => 2,
  "might" => 2,
  "a_grainsdreamcomputepy" => 2,
  "such" => 2,
  "0" => 2,
  "which" => 2,
  "_call_awslatestmetadatas" => 2,
  "openstack20120810meta_datajsoninfo" => 2,
  "latestmetadatainfo" => 2,
  "elif" => 2,
  "should" => 2,
  "latestmetadatablockdevicemappinginfo" => 2,
  "latestmetadatablockdevicemappingamiinfo" => 2,
  "_get_dreamcompute_hostinfopath" => 2,
  "d" => 2,
  "x" => 2,
  "v" => 2,
  "nd" => 2,
  "recursive" => 2,
  "does" => 2,
  "like" => 2,
  "returns" => 2,
  "_snake_caseify_dictdata" => 2,
  "latestmetadatablockdevicemappingebs0info" => 2,
  "try" => 2,
  "_get_dreamcompute_additional" => 2,
  "latestmetadatablockdevicemappingebs1info" => 2,
  "latestmetadatablockdevicemappingrootinfo" => 2,
  "latestmetadatahostnameinfo" => 2,
  "error" => 2,
  "latestmetadatainstanceactioninfo" => 2,
  "ioerror" => 2,
  "dreamcompute_info" => 2,
  "stack" => 2,
}
`;

exports[`extractors When adding stopWords 1`] = `
Object {
  "keywords": Array [
    "api",
    "the",
    "data",
    "that",
    "in",
    "return",
  ],
}
`;
